salt/_modules/better_kubernetes.py

```diff
@@ -0,0 +1,148 @@
+"""Execution methods for management of Kubernetes objects.
+
+This module relies on the `kubernetes` custom Salt utils module for parsing
+K8s object manifests, and providing direct bindings to the Python `kubernetes`
+client.
+
+TODO:
+- handle custom objects
+- add some logging for debug purposes
+- add some documentation for each method
+- support giving a `path` to a file instead of directly the `manifest` data
+- improve management of `kubeconfig` and `context` parameters, relying on
+  master configuration and sane defaults - look into `__opts__`
+"""
+from salt.exceptions import CommandExecutionError
+
+try:
+    import kubernetes.client as k8s_client
+except ImportError:
+    HAS_LIBS = False
+else:
+    HAS_LIBS = True
+
+__virtualname__ = 'metalk8s_kubernetes'
+
+
+def __virtual__():
+    if not HAS_LIBS:
+        return False, 'Missing Python `kubernetes.client` dependency'
+
+    return __virtualname__
+
+
+def _extract_obj_and_details(manifest):
+    try:
+        details = __utils__['kubernetes.get_object_details'](manifest)
+        obj = __utils__['kubernetes.convert_manifest_into_object'](manifest)
+    except ValueError as exc:
+        raise CommandExecutionError('Invalid manifest: {!s}'.format(exc))
+
+    return obj, details
+
+
+def _extract_kwargs(obj, details):
+    kwargs = {'name': obj.metadata.name}
+
+    if details.scope == 'namespaced':
+        kwargs['namespace'] = obj.metadata.namespace
+
+    if hasattr(details, 'group'):
+        # This object represents a CustomResource
+        kwargs.update({
+            'group': details.group,
+            'version': details.version,
+            'plural': details.plural,
+        })
+
+    return kwargs
+
+
+def get_object(manifest, kubeconfig, context):
+    obj, details = _extract_obj_and_details(manifest)
+
+    kwargs = _extract_kwargs(obj, details)
+
+    try:
+        result = details.retrieve(**kwargs)
+    except Exception as exc:  # TODO: which Exception could be raised, and why?
+        raise CommandExecutionError(
+            'Failed to retrieve object: {!s}'.format(exc)
+        )
+
+    if hasattr(result, 'to_dict'):
```

Comment:
Maybe put a comment in which case this attribute *doesn't* exist, and why we're sure `result` is in that case a valid `dict` already.

Reply:
@NicolasT Yes, similar to the special case for `kwargs` with custom objects, I'd like to remove this check altogether. We know that all models for standard objects have a `to_dict` method, so I'll just have something similar for custom objects.

@slaperche-scality Oopsie, forgot that one! Anyway, as explained above, I'll remove the check altogether.